pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,338 @@
+				allocateResult[deviceType] = deviceAllocations
+				return nil
+			}
+		}
+		klog.V(5).Infof("node resource doest not satisfy pod's multiple %v request, expect %v, got %v", deviceType, commonDeviceWanted, satisfiedDeviceCount)
+		return fmt.Errorf("node does not have enough %v", deviceType)
+	}
+
+	for minor, resources := range n.deviceFree[deviceType] {
+		if satisfied, _ := quotav1.LessThanOrEqual(podRequest, resources); satisfied {
+			deviceAllocations = append(deviceAllocations, &apiext.DeviceAllocation{
+				Minor:     int32(minor),
+				Resources: podRequest,
+			})
+			allocateResult[deviceType] = deviceAllocations
+			return nil
+		}
+	}
+	klog.V(5).Infof("node resource doest not satisfy pod's %v request", deviceType)
+	return fmt.Errorf("node does not have enough %v", deviceType)
+}
+
+func (n *nodeDevice) tryAllocateGPU(podRequest corev1.ResourceList, allocateResult apiext.DeviceAllocations) error {
+	podRequest = quotav1.Mask(podRequest, deviceResourceNames[schedulingv1alpha1.GPU])
+	nodeDeviceTotal := n.deviceTotal[schedulingv1alpha1.GPU]
+	if len(nodeDeviceTotal) <= 0 {
+		return fmt.Errorf("node does not have enough GPU")
+	}
+
+	var activeMinor int
+	for i := range nodeDeviceTotal {
+		activeMinor = i
+		break
+	}
+
+	if gpuMem, ok := podRequest[apiext.GPUMemory]; ok {
+		podRequest[apiext.GPUMemoryRatio] = memBytesToRatio(gpuMem, nodeDeviceTotal[activeMinor][apiext.GPUMemory])
+	} else {
+		gpuMemRatio := podRequest[apiext.GPUMemoryRatio]
+		podRequest[apiext.GPUMemory] = memRatioToBytes(gpuMemRatio, nodeDeviceTotal[activeMinor][apiext.GPUMemory])
+	}
+
+	var deviceAllocations []*apiext.DeviceAllocation
+	if isMultipleGPUPod(podRequest) {
+		gpuCore, gpuMem, gpuMemRatio := podRequest[apiext.GPUCore], podRequest[apiext.GPUMemory], podRequest[apiext.GPUMemoryRatio]
+		gpuWanted := gpuCore.Value() / 100
+		podRequestPerCard := corev1.ResourceList{
+			apiext.GPUCore:        *resource.NewQuantity(gpuCore.Value()/gpuWanted, resource.DecimalSI),
+			apiext.GPUMemory:      *resource.NewQuantity(gpuMem.Value()/gpuWanted, resource.BinarySI),
+			apiext.GPUMemoryRatio: *resource.NewQuantity(gpuMemRatio.Value()/gpuWanted, resource.DecimalSI),
+		}
+		satisfiedDeviceCount := 0
+		for minor, resources := range n.deviceFree[schedulingv1alpha1.GPU] {
+			if satisfied, _ := quotav1.LessThanOrEqual(podRequestPerCard, resources); satisfied {
+				satisfiedDeviceCount++
+				deviceAllocations = append(deviceAllocations, &apiext.DeviceAllocation{
+					Minor:     int32(minor),
+					Resources: podRequestPerCard,
+				})
+			}
+			if satisfiedDeviceCount == int(gpuWanted) {
+				allocateResult[schedulingv1alpha1.GPU] = deviceAllocations
+				return nil
+			}
+		}
+		klog.V(5).Infof("node GPU resource doest not satisfy pod's multiple GPU request, expect %v, got %v", gpuWanted, satisfiedDeviceCount)
+		return fmt.Errorf("node does not have enough GPU")
+	}
+	for minor, resources := range n.deviceFree[schedulingv1alpha1.GPU] {
```

Comment:
need  a "else" here?  for single card allocation, compared to multiple cards above.